File: window/window/src/main/java/androidx/window/SidecarCompat.java
/*
 * Copyright 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.window;

import static androidx.window.ActivityUtil.getActivityWindowToken;
import static androidx.window.ExtensionCompat.DEBUG;
import static androidx.window.Version.VERSION_0_1;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.graphics.Rect;
import android.os.IBinder;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;

import androidx.annotation.GuardedBy;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import androidx.collection.SimpleArrayMap;
import androidx.window.sidecar.SidecarDeviceState;
import androidx.window.sidecar.SidecarDisplayFeature;
import androidx.window.sidecar.SidecarInterface;
import androidx.window.sidecar.SidecarProvider;
import androidx.window.sidecar.SidecarWindowLayoutInfo;

import java.lang.ref.WeakReference;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.WeakHashMap;

/** Extension interface compatibility wrapper for v0.1 sidecar. */
@SuppressWarnings("deprecation")
final class SidecarCompat implements ExtensionInterfaceCompat {
    private static final String TAG = "SidecarCompat";

    // Map of active listeners registered with #onWindowLayoutChangeListenerAdded() and not yet
    // removed by #onWindowLayoutChangeListenerRemoved().
    protected final SimpleArrayMap<IBinder, Activity> mWindowListenerRegisteredContexts =
            new SimpleArrayMap<>();

    private ExtensionCallbackInterface mExtensionCallback;
    private final SidecarAdapter mSidecarAdapter;

    @VisibleForTesting
    final SidecarInterface mSidecar;

    SidecarCompat(Context context) {
        this(SidecarProvider.getSidecarImpl(context), new SidecarAdapter());
        if (mSidecar == null) {
            throw new IllegalArgumentException("Sidecar provider returned null");
        }
    }

    @VisibleForTesting
    SidecarCompat(@NonNull SidecarInterface sidecar, SidecarAdapter sidecarAdapter) {
        // Empty implementation to avoid null checks.
        mExtensionCallback = new ExtensionCallbackInterface() {
            @Override
            public void onDeviceStateChanged(@NonNull DeviceState newDeviceState) {

            }

            @Override
            public void onWindowLayoutChanged(@NonNull Activity activity,
                    @NonNull WindowLayoutInfo newLayout) {

            }
        };
        mSidecar = sidecar;
        mSidecarAdapter = sidecarAdapter;
    }

    @Override
    public void setExtensionCallback(@NonNull ExtensionCallbackInterface extensionCallback) {
        mExtensionCallback = new DistinctElementCallback(extensionCallback);
        mSidecar.setSidecarCallback(new DistinctSidecarElementCallback(mSidecarAdapter,
                new TranslatingCallback()));
    }

    //TODO(b/173739071) reduce visibility to @VisibleForTesting
    @NonNull
    WindowLayoutInfo getWindowLayoutInfo(@NonNull Activity activity) {
        IBinder windowToken = getActivityWindowToken(activity);

        SidecarWindowLayoutInfo windowLayoutInfo = mSidecar.getWindowLayoutInfo(windowToken);
        return mSidecarAdapter.translate(activity, windowLayoutInfo, mSidecar.getDeviceState());
    }

    @Override
    public void onWindowLayoutChangeListenerAdded(@NonNull Activity activity) {
        IBinder windowToken = getActivityWindowToken(activity);

        if (windowToken != null) {
            register(windowToken, activity);
        } else {
            FirstAttachAdapter attachAdapter = new FirstAttachAdapter(this, activity);
            activity.getWindow().getDecorView().addOnAttachStateChangeListener(attachAdapter);
        }
    }

    /**
     * Register an {@link IBinder} token and an {@link Activity} so that the given
     * {@link Activity} will receive updates when there is a new {@link WindowLayoutInfo}.
     * @param windowToken for the given {@link Activity}.
     * @param activity that is listening for changes of {@link WindowLayoutInfo}
     */
    void register(@NonNull IBinder windowToken, @NonNull Activity activity) {
        mWindowListenerRegisteredContexts.put(windowToken, activity);
        mSidecar.onWindowLayoutChangeListenerAdded(windowToken);
        mExtensionCallback.onWindowLayoutChanged(activity, getWindowLayoutInfo(activity));
    }

    @Override
    public void onWindowLayoutChangeListenerRemoved(@NonNull Activity activity) {
        IBinder windowToken = getActivityWindowToken(activity);

        mSidecar.onWindowLayoutChangeListenerRemoved(windowToken);

        mWindowListenerRegisteredContexts.remove(windowToken);
    }

    @Override
    public void onDeviceStateListenersChanged(boolean isEmpty) {
        if (!isEmpty) {
            SidecarDeviceState deviceState = mSidecar.getDeviceState();
            mExtensionCallback.onDeviceStateChanged(mSidecarAdapter.translate(deviceState));
        }
        mSidecar.onDeviceStateListenersChanged(isEmpty);
    }

    @SuppressLint("BanUncheckedReflection")
    @Override
    @SuppressWarnings("unused")
    public boolean validateExtensionInterface() {
        try {
            // sidecar.setSidecarCallback(SidecarInterface.SidecarCallback);
            Method methodSetSidecarCallback = mSidecar.getClass().getMethod("setSidecarCallback",
                    SidecarInterface.SidecarCallback.class);
            Class<?> rSetSidecarCallback = methodSetSidecarCallback.getReturnType();
            if (!rSetSidecarCallback.equals(void.class)) {
                throw new NoSuchMethodException("Illegal return type for 'setSidecarCallback': "
                        + rSetSidecarCallback);
            }

            // DO NOT REMOVE SINCE THIS IS VALIDATING THE INTERFACE.
            // sidecar.getDeviceState()
            SidecarDeviceState tmpDeviceState = mSidecar.getDeviceState();

            // sidecar.onDeviceStateListenersChanged(boolean);
            mSidecar.onDeviceStateListenersChanged(true /* isEmpty */);

            // sidecar.getWindowLayoutInfo(IBinder)
            Method methodGetWindowLayoutInfo = mSidecar.getClass()
                    .getMethod("getWindowLayoutInfo", IBinder.class);
            Class<?> rtGetWindowLayoutInfo = methodGetWindowLayoutInfo.getReturnType();
            if (!rtGetWindowLayoutInfo.equals(SidecarWindowLayoutInfo.class)) {
                throw new NoSuchMethodException(
                        "Illegal return type for 'getWindowLayoutInfo': "
                                + rtGetWindowLayoutInfo);
            }

            // sidecar.onWindowLayoutChangeListenerAdded(IBinder);
            Method methodRegisterWindowLayoutChangeListener = mSidecar.getClass()
                    .getMethod("onWindowLayoutChangeListenerAdded", IBinder.class);
            Class<?> rtRegisterWindowLayoutChangeListener =
                    methodRegisterWindowLayoutChangeListener.getReturnType();
            if (!rtRegisterWindowLayoutChangeListener.equals(void.class)) {
                throw new NoSuchMethodException(
                        "Illegal return type for 'onWindowLayoutChangeListenerAdded': "
                                + rtRegisterWindowLayoutChangeListener);
            }

            // sidecar.onWindowLayoutChangeListenerRemoved(IBinder);
            Method methodUnregisterWindowLayoutChangeListener = mSidecar.getClass()
                    .getMethod("onWindowLayoutChangeListenerRemoved", IBinder.class);
            Class<?> rtUnregisterWindowLayoutChangeListener =
                    methodUnregisterWindowLayoutChangeListener.getReturnType();
            if (!rtUnregisterWindowLayoutChangeListener.equals(void.class)) {
                throw new NoSuchMethodException(
                        "Illegal return type for 'onWindowLayoutChangeListenerRemoved': "
                                + rtUnregisterWindowLayoutChangeListener);
            }

            // SidecarDeviceState constructor
            tmpDeviceState = new SidecarDeviceState();

            // deviceState.posture
            // TODO(b/172620880): Workaround for Sidecar API implementation issue.
            try {
                tmpDeviceState.posture = SidecarDeviceState.POSTURE_OPENED;
            } catch (NoSuchFieldError error) {
                if (DEBUG) {
                    Log.w(TAG, "Sidecar implementation doesn't conform to primary interface "
                            + "version, continue to check for the secondary one "
                            + VERSION_0_1 + ", error: " + error);
                }
                Method methodSetPosture = SidecarDeviceState.class.getMethod("setPosture",
                        int.class);
                methodSetPosture.invoke(tmpDeviceState, SidecarDeviceState.POSTURE_OPENED);
                Method methodGetPosture = SidecarDeviceState.class.getMethod("getPosture");
                int posture = (int) methodGetPosture.invoke(tmpDeviceState);
                if (posture != SidecarDeviceState.POSTURE_OPENED) {
                    throw new Exception("Invalid device posture getter/setter");
                }
            }

            // SidecarDisplayFeature constructor
            SidecarDisplayFeature displayFeature = new SidecarDisplayFeature();

            // displayFeature.getRect()/setRect()
            Rect tmpRect = displayFeature.getRect();
            displayFeature.setRect(tmpRect);

            // displayFeature.getType()/setType()
            int tmpType = displayFeature.getType();
            displayFeature.setType(SidecarDisplayFeature.TYPE_FOLD);

            // SidecarWindowLayoutInfo constructor
            SidecarWindowLayoutInfo windowLayoutInfo = new SidecarWindowLayoutInfo();

            // windowLayoutInfo.displayFeatures
            try {
                final List<SidecarDisplayFeature> tmpDisplayFeatures =
                        windowLayoutInfo.displayFeatures;
                // TODO(b/172620880): Workaround for Sidecar API implementation issue.
            } catch (NoSuchFieldError error) {
                if (DEBUG) {
                    Log.w(TAG, "Sidecar implementation doesn't conform to primary interface "
                            + "version, continue to check for the secondary one "
                            + VERSION_0_1 + ", error: " + error);
                }
                List<SidecarDisplayFeature> featureList = new ArrayList<>();
                featureList.add(displayFeature);
                Method methodSetFeatures = SidecarWindowLayoutInfo.class.getMethod(
                        "setDisplayFeatures", List.class);
                methodSetFeatures.invoke(windowLayoutInfo, featureList);
                Method methodGetFeatures = SidecarWindowLayoutInfo.class.getMethod(
                        "getDisplayFeatures");
                @SuppressWarnings("unchecked")
                final List<SidecarDisplayFeature> resultDisplayFeatures =
                        (List<SidecarDisplayFeature>) methodGetFeatures.invoke(windowLayoutInfo);
                if (!featureList.equals(resultDisplayFeatures)) {
                    throw new Exception("Invalid display feature getter/setter");
                }
            }

            return true;
        } catch (Throwable t) {
            if (DEBUG) {
                Log.e(TAG, "Sidecar implementation doesn't conform to interface version "
                        + VERSION_0_1 + ", error: " + t);
            }
            return false;
        }
    }

    @Nullable
    static Version getSidecarVersion() {
        try {
            String vendorVersion = SidecarProvider.getApiVersion();
            return !TextUtils.isEmpty(vendorVersion) ? Version.parse(vendorVersion) : null;
        } catch (NoClassDefFoundError e) {
            if (DEBUG) {
                Log.d(TAG, "Sidecar version not found");
            }
            return null;
        } catch (UnsupportedOperationException e) {
            if (DEBUG) {
                Log.d(TAG, "Stub Sidecar");
            }
            return null;
        }
    }

    /**
     * An adapter that will run a callback when a window is attached and then be removed from the
     * listener set.
     */
    private static class FirstAttachAdapter implements View.OnAttachStateChangeListener {

        private final SidecarCompat mSidecarCompat;
        private final WeakReference<Activity> mActivityWeakReference;

        FirstAttachAdapter(SidecarCompat sidecarCompat, Activity activity) {
            mSidecarCompat = sidecarCompat;
            mActivityWeakReference = new WeakReference<>(activity);
        }

        @Override
        public void onViewAttachedToWindow(View view) {
            view.removeOnAttachStateChangeListener(this);
            Activity activity = mActivityWeakReference.get();
            IBinder token = getActivityWindowToken(activity);
            if (activity == null) {
                if (DEBUG) {
                    Log.d(TAG, "Unable to register activity since activity is missing");
                }
                return;
            }
            if (token == null) {
                if (DEBUG) {
                    Log.w(TAG, "Unable to register activity since the window token is missing");
                }
                return;
            }
            mSidecarCompat.register(token, activity);
        }

        @Override
        public void onViewDetachedFromWindow(View view) { }
    }

    final class TranslatingCallback implements SidecarInterface.SidecarCallback {
        @Override
        @SuppressLint("SyntheticAccessor")
        public void onDeviceStateChanged(@NonNull SidecarDeviceState newDeviceState) {
            mExtensionCallback.onDeviceStateChanged(mSidecarAdapter.translate(newDeviceState));

            for (int i = 0; i < mWindowListenerRegisteredContexts.size(); i++) {
                Activity activity = mWindowListenerRegisteredContexts.valueAt(i);
                IBinder windowToken = getActivityWindowToken(activity);
                if (windowToken == null) {
                    continue;
                }
                SidecarWindowLayoutInfo layoutInfo = mSidecar.getWindowLayoutInfo(windowToken);
                mExtensionCallback.onWindowLayoutChanged(activity,
                        mSidecarAdapter.translate(activity, layoutInfo, newDeviceState));
            }
        }

        @Override
        @SuppressLint("SyntheticAccessor")
        public void onWindowLayoutChanged(@NonNull IBinder windowToken,
                @NonNull SidecarWindowLayoutInfo newLayout) {
            Activity activity = mWindowListenerRegisteredContexts.get(windowToken);
            if (activity == null) {
                Log.w(TAG, "Unable to resolve activity from window token. Missing a call"
                        + "to #onWindowLayoutChangeListenerAdded()?");
                return;
            }

            mExtensionCallback.onWindowLayoutChanged(activity,
                    mSidecarAdapter.translate(activity, newLayout, mSidecar.getDeviceState()));
        }
    }

    /**
     * A class to record the last calculated values from {@link SidecarInterface} and filter out
     * duplicates. This class uses {@link WindowLayoutInfo} and {@link DeviceState} as opposed to
     * {@link SidecarDeviceState} and {@link SidecarDisplayFeature} since the methods
     * {@link Object#equals(Object)} and {@link Object#hashCode()} may not have been overridden.
     */
    private static final class DistinctElementCallback
            implements ExtensionCallbackInterface {

        private final Object mLock = new Object();
        @GuardedBy("mLock")
        private DeviceState mLastDeviceState;
        /**
         * A map from {@link Activity} to the last computed {@link WindowLayoutInfo} for the
         * given activity. A {@link WeakHashMap} is used to avoid retaining the {@link Activity}.
         */
        @GuardedBy("mLock")
        private final WeakHashMap<Activity, WindowLayoutInfo> mActivityWindowLayoutInfo =
                new WeakHashMap<>();
        private final ExtensionCallbackInterface mCallbackInterface;

        DistinctElementCallback(ExtensionCallbackInterface callbackInterface) {
            mCallbackInterface = callbackInterface;
        }

        @Override
        public void onDeviceStateChanged(@NonNull DeviceState newDeviceState) {
            synchronized (mLock) {
                if (newDeviceState.equals(mLastDeviceState)) {
                    return;
                }
                mLastDeviceState = newDeviceState;
                mCallbackInterface.onDeviceStateChanged(newDeviceState);
            }
        }

        @Override
        public void onWindowLayoutChanged(@NonNull Activity activity,
                @NonNull WindowLayoutInfo newLayout) {
            synchronized (mLock) {
                WindowLayoutInfo lastInfo = mActivityWindowLayoutInfo.get(activity);
                if (newLayout.equals(lastInfo)) {
                    return;
                }
                mActivityWindowLayoutInfo.put(activity, newLayout);
            }
            mCallbackInterface.onWindowLayoutChanged(activity, newLayout);
        }
    }

    /**
     * A class to record the last calculated values from {@link SidecarInterface} and filter out
     * duplicates. This class uses {@link SidecarAdapter} to compute equality since the methods
     * {@link Object#equals(Object)} and {@link Object#hashCode()} may not have been overridden.
     */
    private static final class DistinctSidecarElementCallback
            implements SidecarInterface.SidecarCallback {

        private final Object mLock = new Object();
        @GuardedBy("mLock")
        private SidecarDeviceState mLastDeviceState;
        /**
         * A map from {@link Activity} to the last computed {@link WindowLayoutInfo} for the
         * given activity. A {@link WeakHashMap} is used to avoid retaining the {@link Activity}.
         */
        @GuardedBy("mLock")
        private final WeakHashMap<IBinder, SidecarWindowLayoutInfo> mActivityWindowLayoutInfo =
                new WeakHashMap<>();
        private final SidecarAdapter mSidecarAdapter;
        private final SidecarInterface.SidecarCallback mCallbackInterface;

        DistinctSidecarElementCallback(SidecarAdapter adapter,
                SidecarInterface.SidecarCallback callbackInterface) {
            mSidecarAdapter = adapter;
            mCallbackInterface = callbackInterface;
        }

        @Override
        public void onDeviceStateChanged(@NonNull SidecarDeviceState newDeviceState) {
            synchronized (mLock) {
                if (mSidecarAdapter.isEqualSidecarDeviceState(mLastDeviceState, newDeviceState)) {
                    return;
                }
                mLastDeviceState = newDeviceState;
                mCallbackInterface.onDeviceStateChanged(newDeviceState);
            }
        }

        @Override
        public void onWindowLayoutChanged(@NonNull IBinder token,
                @NonNull SidecarWindowLayoutInfo newLayout) {
            synchronized (mLock) {
                SidecarWindowLayoutInfo lastInfo = mActivityWindowLayoutInfo.get(token);
                if (mSidecarAdapter.isEqualSidecarWindowLayoutInfo(lastInfo, newLayout)) {
                    return;
                }
                mActivityWindowLayoutInfo.put(token, newLayout);
            }
            mCallbackInterface.onWindowLayoutChanged(token, newLayout);
        }
    }
}

File: window/window/src/main/java/androidx/window/layout/adapter/sidecar/SidecarCompat.kt
/*
 * Copyright 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// Sidecar is deprecated but we still need to support it.
@file:Suppress("DEPRECATION")

package androidx.window.layout.adapter.sidecar

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.res.Configuration
import android.os.IBinder
import android.text.TextUtils
import android.util.Log
import android.view.View
import androidx.annotation.GuardedBy
import androidx.annotation.VisibleForTesting
import androidx.core.content.OnConfigurationChangedProvider
import androidx.core.util.Consumer
import androidx.window.core.Version
import androidx.window.core.Version.Companion.parse
import androidx.window.layout.WindowLayoutInfo
import androidx.window.layout.adapter.sidecar.ExtensionInterfaceCompat.ExtensionCallbackInterface
import androidx.window.layout.adapter.sidecar.SidecarWindowBackend.Companion.DEBUG
import androidx.window.sidecar.SidecarDeviceState
import androidx.window.sidecar.SidecarDisplayFeature
import androidx.window.sidecar.SidecarInterface
import androidx.window.sidecar.SidecarInterface.SidecarCallback
import androidx.window.sidecar.SidecarProvider
import androidx.window.sidecar.SidecarWindowLayoutInfo
import java.lang.ref.WeakReference
import java.util.WeakHashMap
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/** Extension interface compatibility wrapper for v0.1 sidecar.  */
internal class SidecarCompat @VisibleForTesting constructor(
    @get:VisibleForTesting
    val sidecar: SidecarInterface?,
    private val sidecarAdapter: SidecarAdapter
) : ExtensionInterfaceCompat {
    // Map of active listeners registered with #onWindowLayoutChangeListenerAdded() and not yet
    // removed by #onWindowLayoutChangeListenerRemoved().
    private val windowListenerRegisteredContexts = mutableMapOf<IBinder, Activity>()
    // Map of activities registered to their component callbacks so we can keep track and
    // remove when the activity is unregistered
    private val componentCallbackMap = mutableMapOf<Activity, Consumer<Configuration>>()
    private var extensionCallback: DistinctElementCallback? = null

    constructor(context: Context) : this(
        getSidecarCompat(context),
        SidecarAdapter()
    )

    override fun setExtensionCallback(extensionCallback: ExtensionCallbackInterface) {
        this.extensionCallback = DistinctElementCallback(extensionCallback)
        sidecar?.setSidecarCallback(
            DistinctElementSidecarCallback(
                sidecarAdapter,
                TranslatingCallback()
            )
        )
    }

    @VisibleForTesting
    fun getWindowLayoutInfo(activity: Activity): WindowLayoutInfo {
        val windowToken = getActivityWindowToken(activity) ?: return WindowLayoutInfo(emptyList())
        val windowLayoutInfo = sidecar?.getWindowLayoutInfo(windowToken)
        return sidecarAdapter.translate(
            windowLayoutInfo,
            sidecar?.deviceState ?: SidecarDeviceState()
        )
    }

    override fun onWindowLayoutChangeListenerAdded(activity: Activity) {
        val windowToken = getActivityWindowToken(activity)
        if (windowToken != null) {
            register(windowToken, activity)
        } else {
            val attachAdapter = FirstAttachAdapter(this, activity)
            activity.window.decorView.addOnAttachStateChangeListener(attachAdapter)
        }
    }

    /**
     * Register an [IBinder] token and an [Activity] so that the given
     * [Activity] will receive updates when there is a new [WindowLayoutInfo].
     * @param windowToken for the given [Activity].
     * @param activity that is listening for changes of [WindowLayoutInfo]
     */
    fun register(windowToken: IBinder, activity: Activity) {
        windowListenerRegisteredContexts[windowToken] = activity
        sidecar?.onWindowLayoutChangeListenerAdded(windowToken)
        // Since SidecarDeviceState and SidecarWindowLayout are merged we trigger both
        // data streams.
        if (windowListenerRegisteredContexts.size == 1) {
            sidecar?.onDeviceStateListenersChanged(false)
        }
        extensionCallback?.onWindowLayoutChanged(activity, getWindowLayoutInfo(activity))
        registerConfigurationChangeListener(activity)
    }

    private fun registerConfigurationChangeListener(activity: Activity) {
        // Only register a component callback if we haven't already as register
        // may be called multiple times for the same activity
        if (componentCallbackMap[activity] == null && activity is OnConfigurationChangedProvider) {
            // Create a configuration change observer to send updated WindowLayoutInfo
            // when the configuration of the app changes: b/186647126
            val configChangeObserver = Consumer<Configuration> {
                extensionCallback?.onWindowLayoutChanged(
                    activity,
                    getWindowLayoutInfo(activity)
                )
            }
            componentCallbackMap[activity] = configChangeObserver
            activity.addOnConfigurationChangedListener(configChangeObserver)
        }
    }

    override fun onWindowLayoutChangeListenerRemoved(activity: Activity) {
        val windowToken = getActivityWindowToken(activity) ?: return
        sidecar?.onWindowLayoutChangeListenerRemoved(windowToken)
        unregisterComponentCallback(activity)
        extensionCallback?.clearWindowLayoutInfo(activity)
        val isLast = windowListenerRegisteredContexts.size == 1
        windowListenerRegisteredContexts.remove(windowToken)
        if (isLast) {
            sidecar?.onDeviceStateListenersChanged(true)
        }
    }

    private fun unregisterComponentCallback(activity: Activity) {
        val configChangeObserver = componentCallbackMap[activity] ?: return
        if (activity is OnConfigurationChangedProvider) {
            activity.removeOnConfigurationChangedListener(configChangeObserver)
        }
        componentCallbackMap.remove(activity)
    }

    @SuppressLint("BanUncheckedReflection")
    override fun validateExtensionInterface(): Boolean {
        return try {
            // sidecar.setSidecarCallback(SidecarInterface.SidecarCallback);
            val methodSetSidecarCallback = sidecar?.javaClass?.getMethod(
                "setSidecarCallback",
                SidecarCallback::class.java
            )
            val rSetSidecarCallback = methodSetSidecarCallback?.returnType
            if (rSetSidecarCallback != Void.TYPE) {
                throw NoSuchMethodException(
                    "Illegal return type for 'setSidecarCallback': $rSetSidecarCallback"
                )
            }

            // DO NOT REMOVE SINCE THIS IS VALIDATING THE INTERFACE.
            // sidecar.getDeviceState()
            @Suppress("VARIABLE_WITH_REDUNDANT_INITIALIZER")
            var tmpDeviceState = sidecar?.deviceState

            // sidecar.onDeviceStateListenersChanged(boolean);
            sidecar?.onDeviceStateListenersChanged(true /* isEmpty */)

            // sidecar.getWindowLayoutInfo(IBinder)
            val methodGetWindowLayoutInfo = sidecar?.javaClass
                ?.getMethod("getWindowLayoutInfo", IBinder::class.java)
            val rtGetWindowLayoutInfo = methodGetWindowLayoutInfo?.returnType
            if (rtGetWindowLayoutInfo != SidecarWindowLayoutInfo::class.java) {
                throw NoSuchMethodException(
                    "Illegal return type for 'getWindowLayoutInfo': $rtGetWindowLayoutInfo"
                )
            }

            // sidecar.onWindowLayoutChangeListenerAdded(IBinder);
            val methodRegisterWindowLayoutChangeListener = sidecar?.javaClass
                ?.getMethod("onWindowLayoutChangeListenerAdded", IBinder::class.java)
            val rtRegisterWindowLayoutChangeListener =
                methodRegisterWindowLayoutChangeListener?.returnType
            if (rtRegisterWindowLayoutChangeListener != Void.TYPE) {
                throw NoSuchMethodException(
                    "Illegal return type for 'onWindowLayoutChangeListenerAdded': " +
                        "$rtRegisterWindowLayoutChangeListener"
                )
            }

            // sidecar.onWindowLayoutChangeListenerRemoved(IBinder);
            val methodUnregisterWindowLayoutChangeListener = sidecar?.javaClass
                ?.getMethod("onWindowLayoutChangeListenerRemoved", IBinder::class.java)
            val rtUnregisterWindowLayoutChangeListener =
                methodUnregisterWindowLayoutChangeListener?.returnType
            if (rtUnregisterWindowLayoutChangeListener != Void.TYPE) {
                throw NoSuchMethodException(
                    "Illegal return type for 'onWindowLayoutChangeListenerRemoved': " +
                        "$rtUnregisterWindowLayoutChangeListener"
                )
            }

            // SidecarDeviceState constructor
            tmpDeviceState = SidecarDeviceState()

            // deviceState.posture
            // TODO(b/172620880): Workaround for Sidecar API implementation issue.
            try {
                tmpDeviceState.posture = SidecarDeviceState.POSTURE_OPENED
            } catch (error: NoSuchFieldError) {
                if (DEBUG) {
                    Log.w(
                        TAG,
                        "Sidecar implementation doesn't conform to primary interface version, " +
                            "continue to check for the secondary one ${Version.VERSION_0_1}, " +
                            "error: $error"
                    )
                }
                val methodSetPosture = SidecarDeviceState::class.java.getMethod(
                    "setPosture",
                    Int::class.javaPrimitiveType
                )
                methodSetPosture.invoke(tmpDeviceState, SidecarDeviceState.POSTURE_OPENED)
                val methodGetPosture = SidecarDeviceState::class.java.getMethod("getPosture")
                val posture = methodGetPosture.invoke(tmpDeviceState) as Int
                if (posture != SidecarDeviceState.POSTURE_OPENED) {
                    throw Exception("Invalid device posture getter/setter")
                }
            }

            // SidecarDisplayFeature constructor
            val displayFeature = SidecarDisplayFeature()

            // displayFeature.getRect()/setRect()
            val tmpRect = displayFeature.rect
            displayFeature.rect = tmpRect

            // displayFeature.getType()/setType()
            @Suppress("UNUSED_VARIABLE")
            val tmpType = displayFeature.type
            displayFeature.type = SidecarDisplayFeature.TYPE_FOLD

            // SidecarWindowLayoutInfo constructor
            val windowLayoutInfo = SidecarWindowLayoutInfo()

            // windowLayoutInfo.displayFeatures
            try {
                @Suppress("UNUSED_VARIABLE")
                val tmpDisplayFeatures = windowLayoutInfo.displayFeatures
                // TODO(b/172620880): Workaround for Sidecar API implementation issue.
            } catch (error: NoSuchFieldError) {
                if (DEBUG) {
                    Log.w(
                        TAG,
                        "Sidecar implementation doesn't conform to primary interface version, " +
                            "continue to check for the secondary one ${Version.VERSION_0_1}, " +
                            "error: $error"
                    )
                }
                val featureList: MutableList<SidecarDisplayFeature> = ArrayList()
                featureList.add(displayFeature)
                val methodSetFeatures = SidecarWindowLayoutInfo::class.java.getMethod(
                    "setDisplayFeatures", MutableList::class.java
                )
                methodSetFeatures.invoke(windowLayoutInfo, featureList)
                val methodGetFeatures = SidecarWindowLayoutInfo::class.java.getMethod(
                    "getDisplayFeatures"
                )
                @Suppress("UNCHECKED_CAST")
                val resultDisplayFeatures =
                    methodGetFeatures.invoke(windowLayoutInfo) as List<SidecarDisplayFeature>
                if (featureList != resultDisplayFeatures) {
                    throw Exception("Invalid display feature getter/setter")
                }
            }
            true
        } catch (t: Throwable) {
            if (DEBUG) {
                Log.e(
                    TAG,
                    "Sidecar implementation doesn't conform to interface version " +
                        "${Version.VERSION_0_1}, error: $t"
                )
            }
            false
        }
    }

    /**
     * An adapter that will run a callback when a window is attached and then be removed from the
     * listener set.
     */
    private class FirstAttachAdapter(
        private val sidecarCompat: SidecarCompat,
        activity: Activity
    ) : View.OnAttachStateChangeListener {
        private val activityWeakReference = WeakReference(activity)
        override fun onViewAttachedToWindow(view: View) {
            view.removeOnAttachStateChangeListener(this)
            val activity = activityWeakReference.get()
            val token = getActivityWindowToken(activity)
            if (activity == null) {
                if (DEBUG) {
                    Log.d(TAG, "Unable to register activity since activity is missing")
                }
                return
            }
            if (token == null) {
                if (DEBUG) {
                    Log.w(TAG, "Unable to register activity since the window token is missing")
                }
                return
            }
            sidecarCompat.register(token, activity)
        }

        override fun onViewDetachedFromWindow(view: View) {}
    }

    /**
     * A callback to translate from Sidecar classes to local classes.
     *
     * If you change the name of this class, you must update the proguard file.
     */
    internal inner class TranslatingCallback : SidecarCallback {
        override fun onDeviceStateChanged(newDeviceState: SidecarDeviceState) {
            windowListenerRegisteredContexts.values.forEach { activity ->
                val layoutInfo = getActivityWindowToken(activity)
                    ?.let { windowToken -> sidecar?.getWindowLayoutInfo(windowToken) }
                extensionCallback?.onWindowLayoutChanged(
                    activity,
                    sidecarAdapter.translate(layoutInfo, newDeviceState)
                )
            }
        }

        override fun onWindowLayoutChanged(
            windowToken: IBinder,
            newLayout: SidecarWindowLayoutInfo
        ) {
            val activity = windowListenerRegisteredContexts[windowToken]
            if (activity == null) {
                Log.w(
                    TAG,
                    "Unable to resolve activity from window token. Missing a call to " +
                        "#onWindowLayoutChangeListenerAdded()?"
                )
                return
            }
            val layoutInfo = sidecarAdapter.translate(
                newLayout,
                sidecar?.deviceState ?: SidecarDeviceState()
            )
            extensionCallback?.onWindowLayoutChanged(activity, layoutInfo)
        }
    }

    /**
     * A class to record the last calculated values from [SidecarInterface] and filter out
     * duplicates. This class uses [WindowLayoutInfo] as opposed to
     * [SidecarDisplayFeature] since the methods [Object.equals] and
     * [Object.hashCode] may not have been overridden.
     */
    private class DistinctElementCallback(
        private val callbackInterface: ExtensionCallbackInterface
    ) : ExtensionCallbackInterface {
        private val globalLock = ReentrantLock()

        /**
         * A map from [Activity] to the last computed [WindowLayoutInfo] for the
         * given activity. A [WeakHashMap] is used to avoid retaining the [Activity].
         */
        @GuardedBy("globalLock")
        private val activityWindowLayoutInfo = WeakHashMap<Activity, WindowLayoutInfo>()
        override fun onWindowLayoutChanged(
            activity: Activity,
            newLayout: WindowLayoutInfo
        ) {
            globalLock.withLock {
                val lastInfo = activityWindowLayoutInfo[activity]
                if (newLayout == lastInfo) {
                    return
                }
                activityWindowLayoutInfo.put(activity, newLayout)
            }
            callbackInterface.onWindowLayoutChanged(activity, newLayout)
        }

        fun clearWindowLayoutInfo(activity: Activity) {
            globalLock.withLock {
                activityWindowLayoutInfo[activity] = null
            }
        }
    }

    companion object {
        private const val TAG = "SidecarCompat"
        val sidecarVersion: Version?
            get() = try {
                val vendorVersion = SidecarProvider.getApiVersion()
                if (!TextUtils.isEmpty(vendorVersion)) parse(vendorVersion) else null
            } catch (e: NoClassDefFoundError) {
                if (DEBUG) {
                    Log.d(TAG, "Sidecar version not found")
                }
                null
            } catch (e: UnsupportedOperationException) {
                if (DEBUG) {
                    Log.d(TAG, "Stub Sidecar")
                }
                null
            }

        internal fun getSidecarCompat(context: Context): SidecarInterface? {
            return SidecarProvider.getSidecarImpl(context.applicationContext)
        }

        /**
         * A utility method [Activity] to return an optional [IBinder] window token from an
         * [Activity].
         */
        internal fun getActivityWindowToken(activity: Activity?): IBinder? {
            return activity?.window?.attributes?.token
        }
    }
}
